vpr/src/pack/cluster_legalizer.h

```diff
@@ -0,0 +1,478 @@
+ * that the cluster is fully legalized while it is being created. An example
+ * of how to sue this strategy would look something like this:
+ *
+ * Clusterlegalizer legalizer(...);
+ *
+ * std::tie(pack_result, new_cluster_id) = legalizer.start_new_cluster(seed_mol,
+ *                                                                     cluster_type,
+ *                                                                     mode);
+ * for mol in molecules_to_add:
+ *      // Do the expensive check for each molecule added.
+ *      status = legalizer.add_mol_to_cluster(mol, new_cluster_id);
+ *      if (status != e_block_pack_status::BLK_PASSED)
+ *          break;
+ *
+ * // new_cluster_id now contains a fully legalized cluster.
+ */
+class ClusterLegalizer {
+public:
+    // Iterator for the legalization cluster IDs
+    typedef typename vtr::vector_map<LegalizationClusterId, LegalizationClusterId>::const_iterator cluster_iterator;
+
+    // Range for the legalization cluster IDs
+    typedef typename vtr::Range<cluster_iterator> cluster_range;
+
+private:
+
+    /*
+     * @brief Helper method that tries to pack the given molecule into a cluster.
+     *
+     * This method runs all the legality checks specified by the legalization
+     * strategy. If the molecule can be packed into the cluster, it will insert
+     * it into the cluster.
+     *
+     *  @param molecule                 The molecule to insert into the cluster.
+     *  @param cluster                  The cluster to try to insert the molecule into.
+     *  @param cluster_id               The ID of the cluster.
+     *  @param max_external_pin_util    The max external pin utilization for a
+     *                                  cluster of this type.
+     */
+    e_block_pack_status try_pack_molecule(t_pack_molecule* molecule,
+                                          LegalizationCluster& cluster,
+                                          LegalizationClusterId cluster_id,
+                                          const t_ext_pin_util& max_external_pin_util);
+
+public:
+
+    // Explicitly deleted default constructor. Need to use other constructor to
+    // initialize state correctly.
+    ClusterLegalizer() = delete;
+
+    /*
+     * @brief Initialize the ClusterLegalizer class.
+     *
+     * Allocates internal state.
+     */
+    ClusterLegalizer(const AtomNetlist& atom_netlist,
+                     const Prepacker& prepacker,
+                     const std::vector<t_logical_block_type>& logical_block_types,
+                     std::vector<t_lb_type_rr_node>* lb_type_rr_graphs,
+                     size_t num_models,
+                     const std::vector<std::string>& target_external_pin_util_str,
+                     const t_pack_high_fanout_thresholds& high_fanout_thresholds,
+                     ClusterLegalizationStrategy cluster_legalization_strategy,
+                     bool enable_pin_feasibility_filter,
```

Comment:
I originally avoided this comment since a lot of the arguments I feel are unnecessary and should be cleaned up later (most of them are just to pre-allocate buffers to max values instead of just using dynamic arrays). I added comments for each param anyways and added a TODO to replace the static arrays for dynamic ones. Ill raise an issue after this PR is merged.